File: python/subsystems/shooter.py
# Copyright (c) FIRST and other WPILib contributors.
# Open Source Software; you can modify and/or share it under the terms of
# the WPILib BSD license file in the root directory of this project.

import wpilib
from commands3 import Mechanism, Command, Scheduler
from subsystems.leds import Leds
from constants import ShooterConstants


class Shooter(Mechanism):
    """Shooter mechanism with velocity control and ball feeding."""

    def __init__(self, leds: Leds):
        super().__init__()
        self._leds = leds
        self._last_speed = 0.0

        self._shooter_motor = wpilib.ExpansionHubMotor(1, ShooterConstants.kShooterMotorPort)
        self._left_feeder_servo = wpilib.ExpansionHubServo(1, 0)
        self._right_feeder_servo = wpilib.ExpansionHubServo(1, 2)

        self._shooter_motor.setReversed(True)
        self._shooter_motor.setDistancePerCount(ShooterConstants.kEncoderDistancePerPulse)
        self._shooter_motor.setEnabled(True)

        self._left_feeder_servo.setContinousRotationMode(True)
        self._right_feeder_servo.setContinousRotationMode(True)

        self._left_feeder_servo.setReversed(True)

        self._left_feeder_servo.setEnabled(True)
        self._right_feeder_servo.setEnabled(True)

        pid_constants = self._shooter_motor.getVelocityPidConstants()
        pid_constants.setPID(ShooterConstants.kP, ShooterConstants.kI, ShooterConstants.kD)
        pid_constants.setFF(ShooterConstants.kS, ShooterConstants.kV, ShooterConstants.kA)

        self.setDefaultCommand(
            self.runRepeatedly(lambda: (self.setSpeed(0), self.setFeed(False)))
            .withPriority(Command.LOWEST_PRIORITY)
            .named("Default Shooter")
        )

        Scheduler.getDefault().addPeriodic(self.periodic)

    def periodic(self):
        if self._last_speed == 0:
            self._leds.setAllBlue()
            return
        error = self.getShooterVelocity() - self._last_speed
        if abs(error) > 2:
            self._leds.setAllRed()
        else:
            self._leds.setAllGreen()

    def getShooterVelocity(self):
        return self._shooter_motor.getEncoderVelocity()

    def getShooterPosition(self):
        return self._shooter_motor.getEncoderPosition()

    def getShooterCurrent(self):
        return self._shooter_motor.getCurrent()

    def isHubConnected(self):
        return self._shooter_motor.isHubConnected()

    def setSpeed(self, speed):
        self._last_speed = speed
        self._shooter_motor.setVelocitySetpoint(speed)

    def setFeed(self, feed):
        error = self.getShooterVelocity() - self._last_speed
        if abs(error) > 2:
            feed = False

        if feed:
            self._left_feeder_servo.set(1.0)
            self._right_feeder_servo.set(1.0)
        else:
            self._left_feeder_servo.set(0)
            self._right_feeder_servo.set(0)

    def getSpinCommand(self):
        """Returns a command that spins the shooter at 40 RPM without feeding."""
        return (
            self.runRepeatedly(lambda: (self.setSpeed(40), self.setFeed(False)))
            .withPriority(Command.DEFAULT_PRIORITY)
            .named("Spin Shooter")
        )

    def getSpinAndFeedCommand(self):
        """Returns a command that spins the shooter and feeds balls."""
        return (
            self.runRepeatedly(lambda: (self.setSpeed(40), self.setFeed(True)))
            .withPriority(Command.DEFAULT_PRIORITY + 1)
            .named("Spin and Feed Shooter")
        )

    def shootTime(self, time):
        """Returns a command that spins and feeds for the given number of seconds."""
        def _shoot(c):
            self.setSpeed(40)
            self.setFeed(True)
            c.wait(time)
            self.setSpeed(0)
            self.setFeed(False)

        return (
            self.run(_shoot)
            .withPriority(Command.DEFAULT_PRIORITY)
            .named(f"Shoot {time}")
        )

File: cpp/src/main/cpp/robot/subsystems/Shooter.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Shooter.h"

#include <cmath>
#include <string>

#include <frc/commands3/Command.h>
#include <frc/commands3/Scheduler.h>
#include <units/time.h>

#include "../Constants.h"

using namespace frc::robot::subsystems;

Shooter::Shooter(Leds& leds) : m_leds{leds} {
  m_shooterMotor.SetReversed(true);
  m_shooterMotor.SetDistancePerCount(ShooterConstants::kEncoderDistancePerPulse);
  m_shooterMotor.SetEnabled(true);

  m_leftFeederServo.SetContinuousRotationMode(true);
  m_rightFeederServo.SetContinuousRotationMode(true);

  m_leftFeederServo.SetReversed(true);

  m_leftFeederServo.SetEnabled(true);
  m_rightFeederServo.SetEnabled(true);

  auto pidConstants = m_shooterMotor.GetVelocityPidConstants();
  pidConstants.SetPID(ShooterConstants::kP, ShooterConstants::kI,
                      ShooterConstants::kD);
  pidConstants.SetFF(ShooterConstants::kS, ShooterConstants::kV,
                     ShooterConstants::kA);

  SetDefaultCommand(RunRepeatedly([this] {
                      SetSpeed(0);
                      SetFeed(false);
                    })
                        .WithPriority(frc::commands3::Command::kLowestPriority)
                        .Named("Default Shooter"));

  frc::commands3::Scheduler::GetDefault().AddPeriodic(
      [this] { Periodic(); });
}

void Shooter::Periodic() {
  if (m_lastSpeed == 0.0) {
    m_leds.SetAllBlue();
    return;
  }
  double error = GetShooterVelocity() - m_lastSpeed;
  if (std::abs(error) > 2.0) {
    m_leds.SetAllRed();
  } else {
    m_leds.SetAllGreen();
  }
}

double Shooter::GetShooterVelocity() {
  return m_shooterMotor.GetEncoderVelocity();
}

double Shooter::GetShooterPosition() {
  return m_shooterMotor.GetEncoderPosition();
}

units::ampere_t Shooter::GetShooterCurrent() {
  return m_shooterMotor.GetCurrent();
}

bool Shooter::IsHubConnected() {
  return m_shooterMotor.IsHubConnected();
}

void Shooter::SetSpeed(double speed) {
  m_lastSpeed = speed;
  m_shooterMotor.SetVelocitySetpoint(speed);
}

void Shooter::SetFeed(bool feed) {
  double error = GetShooterVelocity() - m_lastSpeed;
  if (std::abs(error) > 2.0) {
    feed = false;
  }

  if (feed) {
    m_leftFeederServo.Set(1.0);
    m_rightFeederServo.Set(1.0);
  } else {
    m_leftFeederServo.Set(0.0);
    m_rightFeederServo.Set(0.0);
  }
}

frc::commands3::Command Shooter::GetSpinCommand() {
  return RunRepeatedly([this] {
           SetSpeed(40);
           SetFeed(false);
         })
      .WithPriority(frc::commands3::Command::kDefaultPriority)
      .Named("Spin Shooter");
}

frc::commands3::Command Shooter::GetSpinAndFeedCommand() {
  return RunRepeatedly([this] {
           SetSpeed(40);
           SetFeed(true);
         })
      .WithPriority(frc::commands3::Command::kDefaultPriority + 1)
      .Named("Spin and Feed Shooter");
}

frc::commands3::Command Shooter::ShootTime(double time) {
  return Run([this, time](auto& c) {
           SetSpeed(40);
           SetFeed(true);
           c.Wait(units::second_t{time});
           SetSpeed(0);
           SetFeed(false);
         })
      .WithPriority(frc::commands3::Command::kDefaultPriority)
      .Named("Shoot " + std::to_string(time));
}
